dagshub/data_engine/model/datasource.py

```diff
@@ -543,49 +617,111 @@ def _upload_metadata(self, metadata_entries: List[DatapointMetadataUpdateEntry])
         Args:
-            path: Path to save the datasource or dataset file in, defaults to current directory
-            name: Optional: name of the file. if not provided, datasource or dataset name will be used instead.
+            path: Path to the directory where to save the datasource or dataset file.
+                Defaults to current working directory.
+            name: Optional: name of the file. if not provided, datasource name will be used instead.
 
         Returns:
             The path to the saved file
         """
-        res = DatasourceFileStruct(
-            id=self._source.id,
-            repo=self._source.repo,
-            name=self._source.name,
-            datasetQuery=self.serialize_gql_query_input()
-        )
-        file_path = os.path.join(path, (name or res.name) + ".dagshub")
+        file_path = os.path.join(path, (name or self.source.name) + ".json")
+        res = self._to_dict()
         with open(file_path, "w") as file:
-            file.write(json.dumps(res.to_dict(), indent=4))
+            file.write(json.dumps(res, indent=4))
         log_message(f"Datasource saved to '{file_path}'")
 
         return file_path
 
+    def _serialize(self) -> DatasourceSerializedState:
+        res = DatasourceSerializedState(
+            repo=self.source.repo,
+            datasource_id=self.source.id,
+            datasource_name=self.source.name,
+            query=self.serialize_gql_query_input(),
+            timestamp=datetime.datetime.now().timestamp(),
+            modified=self.is_query_different_from_dataset,
+        )
+        if self.assigned_dataset is not None:
+            res.dataset_id = self.assigned_dataset.dataset_id
+            res.dataset_name = self.assigned_dataset.dataset_name
+        return res
+
+    def _to_dict(self) -> Dict:
+        res = self._serialize().to_dict()
+        # Skip Nones in the result
+        res = {k: v for k, v in res.items() if v is not None}
+        return res
+
+    @property
+    def is_query_different_from_dataset(self) -> Optional[bool]:
+        if self.assigned_dataset is None:
+            return None
+        return self._query.to_dict() != self.assigned_dataset.query.to_dict()
+
+    @staticmethod
+    def load_from_serialized_state(state: Dict) -> "Datasource":
+        """
+        Load a Datasource that was saved with :func:`save_to_file`
+
+        Args:
+            state: Serialized JSON object
+        """
+
+        state: DatasourceSerializedState = dacite.from_dict(DatasourceSerializedState, state, config=dacite_config)
+
+        ds_state = DatasourceState(repo=state.repo, name=state.datasource_name, id=state.datasource_id)
+        ds_state.get_from_dagshub()
+        ds = Datasource(ds_state)
+
+        if state.has_query:
+            ds._deserialize_from_gql_result(state.query)
+
+        if state.dataset_id is not None:
+            ds.load_from_dataset(state.dataset_id, state.dataset_name, change_query=False)
+
+        return ds
```

Comment:
I think it's pretty clumsy that you need to use both the `state` and the `ds_state` together to construct one `Datasource` instance, don't you think?
When you pass `ds_state` into the constructor, I kind of expect it to do everything needed

Reply:
ds_state lacks any information about the dataset/query, that's why the constructor has the additional fields aside from `DatasourceState`. Meanwhile `DatasourceSerializedState` has both the Datasource information and Query information.

I guess this goes hand in hand with your complaint about the actual purpose of the `DatasourceState` class. Not sure what I should be doing about it though?

Reply:
I guess I couldn't figure out from reading the code why `DatasourceState` needs to exist apart from `DatasourceSerializedState` but I might have missed the reason.
Is there a case where you need only one and not the other, in both directions?